ClassCastException: with documentLink in catalog XMLCatalogDocumentLinkParticipant

Open the given XML file:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE catalog PUBLIC "-//OASIS//DTD XML Catalogs V1.1//EN" "https://www.oasis-open.org/committees/entity/release/1.1/catalog.dtd" >
<catalog xmlns="urn:oasis:names:tc:entity:xmlns:xml:catalog">
    <!-- W3C schema -->
 
    <uri name="urn:speedata.de:2009/publisher/en" uri="layoutschema-en.xsd"/>
 
</catalog>
```

The document link doesn't work for the uri attribute and you can see the error message:

```
[Error - 18:33:14] Request textDocument/documentLink failed.
  Message: Internal error.
  Code: -32603 
java.util.concurrent.CompletionException: java.lang.ClassCastException: class org.eclipse.lemminx.dom.DOMDocumentType cannot be cast to class org.eclipse.lemminx.dom.DOMElement (org.eclipse.lemminx.dom.DOMDocumentType and org.eclipse.lemminx.dom.DOMElement are in unnamed module of loader 'app')
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1238)
	at java.base/java.util.concurrent.CompletableFuture$BiApply.tryFire(CompletableFuture.java:1205)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Caused by: java.lang.ClassCastException: class org.eclipse.lemminx.dom.DOMDocumentType cannot be cast to class org.eclipse.lemminx.dom.DOMElement (org.eclipse.lemminx.dom.DOMDocumentType and org.eclipse.lemminx.dom.DOMElement are in unnamed module of loader 'app')
	at org.eclipse.lemminx.extensions.catalog.CatalogUtils.getCatalogEntries(CatalogUtils.java:67)
	at org.eclipse.lemminx.extensions.catalog.XMLCatalogDocumentLinkParticipant.findDocumentLinks(XMLCatalogDocumentLinkParticipant.java:36)
	at org.eclipse.lemminx.services.XMLDocumentLink.findDocumentLinks(XMLDocumentLink.java:38)
	at org.eclipse.lemminx.services.XMLLanguageService.findDocumentLinks(XMLLanguageService.java:232)
	at org.eclipse.lemminx.XMLTextDocumentService.lambda$documentLink$12(XMLTextDocumentService.java:363)
	at java.base/java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1236)
	... 7 more
```

If you remove DOCTYPE, it works good.

To fix theproblem we should check the node is an element https://github.com/eclipse/lemminx/blob/dbf43f3cae27a726c22e319eca939872c83fb3d8/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/catalog/CatalogUtils.java#L66

--- a/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/catalog/CatalogUtils.java
+++ b/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/catalog/CatalogUtils.java
@@ -63,7 +63,7 @@ public class CatalogUtils {
 			return Collections.emptyList();
 		}
 		for (DOMNode n : document.getChildren()) {
-			if (CATALOG_ENTITY_NAME.equals(n.getNodeName())) {
+			if (n.isElement() && CATALOG_ENTITY_NAME.equals(n.getNodeName())) {
 				return collectCatalogEntries((DOMElement) n);
 			}
 		}
@@ -178,11 +178,13 @@ public class CatalogUtils {
 	}
 
 	/**
-	 * Returns a catalog entry for the given element and null if a catalog entry can't be made
+	 * Returns a catalog entry for the given element and null if a catalog entry
+	 * can't be made
 	 *
 	 * @param baseURI the base URI of the catalog entry
 	 * @param element the element to turn into a catalog entry
-	 * @return a catalog entry for the given element and null if a catalog entry can't be made
+	 * @return a catalog entry for the given element and null if a catalog entry
+	 *         can't be made
 	 */
 	private static CatalogEntry createCatalogEntry(@NonNull String baseURI, DOMElement element) {
 		if (isCatalogEntryWithURI(element)) {

--- a/org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/catalog/XMLCatalogDocumentLinkTest.java
+++ b/org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/catalog/XMLCatalogDocumentLinkTest.java
@@ -10,11 +10,11 @@
 
 package org.eclipse.lemminx.extensions.catalog;
 
-import org.junit.jupiter.api.Test;
-
-import static org.eclipse.lemminx.XMLAssert.testDocumentLinkFor;
 import static org.eclipse.lemminx.XMLAssert.dl;
 import static org.eclipse.lemminx.XMLAssert.r;
+import static org.eclipse.lemminx.XMLAssert.testDocumentLinkFor;
+
+import org.junit.jupiter.api.Test;
 
 /**
  * Tests for the document links in XML catalog files
@@ -51,6 +51,21 @@ public class XMLCatalogDocumentLinkTest {
 				"</catalog>";
 		testDocumentLinkFor(xml, CATALOG_PATH, //
 				dl(r(3, 9, 3, 23), "src/test/resources/neatSchema.xsd"));
+	}
+
+	@Test
+	public void testCatalogWithCatalogDOCTYPE() {
+		String xml = "<!DOCTYPE catalog PUBLIC \"-//OASIS//DTD XML Catalogs V1.1//EN\" \"catalog.dtd\" >\n"
+				+ //
+				"<catalog xmlns=\"urn:oasis:names:tc:entity:xmlns:xml:catalog\">\n" + //
+				"  <uri\n" + //
+				"    id=\"http://example.org\"\n" + //
+				"    uri=\"neatSchema.xsd\" />\n" + //
+				"</catalog>";
+		testDocumentLinkFor(xml, CATALOG_PATH, //
+				dl(r(0, 64, 0, 75), "src/test/resources/catalog.dtd"), //
+				dl(r(4, 9, 4, 23), "src/test/resources/neatSchema.xsd"));
+
 	}
 
 	@Test

